### ISSUE TITLE
[cc] macro usages highlighted inconsistently

### ISSUE THREAD

**Issue by mtaran-google** [codemirror/google-modes#94]
```cpp
class Foo {
  SOME_MACRO(SomethingThatLooksLikeAType);
};
SOME_OTHER_MACRO(SomethingThatLooksLikeAType);
```
The two macro usages are highlighted differently, and the `SomethingThatLooksLikeAType` are also highlighted differently.

Would it be possible to make it so that UPPER_SNAKE_CASE identifiers are just generally outputted without highlighting, since they're either macros or constants?

**Comment by marijnh** [comment]
The mode tries to parse a grammar, and preprocessor macros, which may appear everywhere and don't follow that grammar, are rather at odds with that. We can add them as special cases for a few situations (expression, statement, class field, maybe?), but this is not something that can be solved generally.

Are the forms that macro invocations can take `NAME` and `NAME(expr, list)` or are there others?

**Comment by mtaran-google** [comment]
Inside a class/struct definition, they'd be NAME(Something), NAME(Something, SomethingElse), etc.

In the same context, it would also be good to properly highlight and more quickly recover from usage of `#include` and similar, e.g.

class Foo {
 public:
  #include "bar/baz.h"
 public:
  int foo();
}
```
currently has no highlighting for the second `public` and the `#include` line itself does not get highlighted as `meta` like other `#include` lines

**Comment by marijnh** [comment]
Patch 230a892 takes care of `#include` lines that are indented (though your indentation test cases seem to require such lines to never be indented, so if you reindent them they'll lose their indentation).

Does upper snake case mean the identifier has at least one underscore? That's what patch fb96191 implements, because we already had some uppercase single-letter variable names in the test cases, which we don't want to match as macros.

But I do think we'll still run into a bunch of problems with this—the class field macros in your example expect a semicolon after them, but you can also define a macro that already includes the semicolon, or expects a block after it, or, as in #100, doesn't make up a full class field, only a part of it. The possibilities are literally endless, so supporting this generally remains impossible.

I'm merging #100 into this, since it's about the same issue.

**Comment by mtaran-google** [comment]
Macros should be in UPPER_SNAKE_CASE (and it should be fine to restrict them to be more than one character long). Per the [style guide](https://google.github.io/styleguide/cppguide.html), only macro names and enum values can be named this way, so I think it should work to consider anything of that format with trailing `(` to be a macro.

**Comment by mtaran-google** [comment]
Styling them as default could work, also possibly meta might be okay? (especially if meta stops being used for namespaces etc. per another bug)

**Comment by marijnh** [comment]
Sure, but that doesn't address the fact that, as in #100, they can take any syntactic role, which a grammar can not encode.

**Comment by mtaran-google** [comment]
Why can't the grammar encode it? Right now you have `Macro` as a possible replacement for `Statement`, but it could just as easily be a possible replacement for any of the other non-terminals, no?

**Comment by marijnh** [comment]
It would still be impossible to know which non-terminal it replaces—it might just replace a type, or the whole rest of the statement, or the type and declarator, and so on. Since macros are string-replacement-based, they might even replace something that doesn't even correspond to a non-terminal in the grammar.

**Comment by mtaran-google** [comment]
> It would still be impossible to know which non-terminal it replaces—it might just replace a type, or the whole rest of the statement, or the type and declarator, and so on.
I can see how this would be a problem in the abstract, but macro usage is generally pretty limited here. I don't expect such conflicts to come up if macros were gradually added to the grammar in the places where they were necessary to fix issues like the one here and in #100. Overall I would not expect there to be more than 5-10 places in the grammar where we actually use macros.

> Since macros are string-replacement-based, they might even replace something that doesn't even correspond to a non-terminal in the grammar.
Sure, but that would definitely be out of scope for this mode (and I don't believe it occurs in our code with any regularity).

**Comment by marijnh** [comment]
Okay, patch 7bc7396 addresses the case of #100, and fb96191 handles the cases from this issue, so I'll close this until the next one is found.
